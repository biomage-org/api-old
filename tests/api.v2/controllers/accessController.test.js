// @ts-nocheck
const userAccessController = require('../../../src/api.v2/controllers/accessController');
const getExperimentUsers = require('../../../src/api.v2/helpers/access/getExperimentUsers');
const createUserInvite = require('../../../src/api.v2/helpers/access/createUserInvite');
const removeAccess = require('../../../src/api.v2/helpers/access/removeAccess');
const postRegistrationHandler = require('../../../src/api.v2/helpers/access/postRegistrationHandler');

const OK = require('../../../src/utils/responses/OK');
const AccessRole = require('../../../src/utils/enums/AccessRole');

jest.mock('../../../src/api.v2/helpers/access/getExperimentUsers');
jest.mock('../../../src/api.v2/helpers/access/createUserInvite');
jest.mock('../../../src/api.v2/helpers/access/removeAccess');
jest.mock('../../../src/api.v2/helpers/access/postRegistrationHandler');

const mockRes = {
  json: jest.fn(),
};

const mockUsersList = [
  {
    name: 'Mock Admin',
    email: 'admin@example.com',
    role: 'admin',
  },
  {
    name: 'Mock User',
    email: 'user@example.com',
    role: 'owner',
  },
];

describe('accessController', () => {
  beforeEach(async () => {
    jest.clearAllMocks();
  });

  it('getExperimentUsers works correctly', async () => {
    const mockReq = { params: { experimentId: 'mockExperimentId' } };

    getExperimentUsers.mockImplementationOnce(
      () => Promise.resolve(mockUsersList),
    );

    await userAccessController.getUserAccess(mockReq, mockRes);

    expect(getExperimentUsers).toHaveBeenCalledWith('mockExperimentId');
    expect(mockRes.json).toHaveBeenCalledWith(mockUsersList);
  });

  it('inviteUser works correctly', async () => {
    const mockReq = {
      user: { email: 'owner@example.com' },
      params: { experimentId: 'mockExperimentId' },
      body: { userEmail: 'user@example.com', role: AccessRole.ADMIN },
    };

    createUserInvite.mockImplementationOnce(
      () => Promise.resolve(),
    );

    await userAccessController.inviteUser(mockReq, mockRes);

    const callParams = createUserInvite.mock.calls[0];

    expect(callParams).toMatchSnapshot();
    expect(mockRes.json).toHaveBeenCalledWith(OK());
  });

  it('revokeAccess works correctly', async () => {
    const mockReq = {
      params: { experimentId: 'mockExperimentId' },
      body: { userEmail: 'user@example.com' },
    };

    removeAccess.mockImplementationOnce(
      () => Promise.resolve(),
    );

    await userAccessController.revokeAccess(mockReq, mockRes);

    const callParams = removeAccess.mock.calls[0];

    expect(callParams).toMatchSnapshot();
    expect(mockRes.json).toHaveBeenCalledWith(OK());
  });

  it('postRegistration works correctly', async () => {
    const mockReq = {
      body: JSON.stringify({ userId: 'mockUserId', userEmail: 'user@example.com' }),
    };

    postRegistrationHandler.mockImplementationOnce(
      () => Promise.resolve(),
    );

    await userAccessController.postRegistration(mockReq, mockRes);

    const callParams = postRegistrationHandler.mock.calls[0];

    expect(callParams).toMatchSnapshot();
    expect(mockRes.json).toHaveBeenCalledWith(OK());
  });
});
