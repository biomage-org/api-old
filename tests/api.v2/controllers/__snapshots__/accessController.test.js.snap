// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accessController inviteUser works correctly 1`] = `
Array [
  "mockExperimentId",
  "user@example.com",
  "admin",
  Object {
    "email": "owner@example.com",
  },
]
`;

exports[`accessController postRegistration works correctly 1`] = `
Array [
  Object {
    "body": "{\\"userId\\":\\"mockUserId\\",\\"userEmail\\":\\"user@example.com\\"}",
  },
]
`;

exports[`accessController revokeAccess works correctly 1`] = `
Array [
  "mockExperimentId",
  "user@example.com",
]
`;
